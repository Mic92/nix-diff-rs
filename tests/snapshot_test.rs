use insta::assert_snapshot;
use std::path::PathBuf;
use std::process::Command;

fn run_nix_diff(file1: &str, file2: &str) -> String {
    let tests_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");

    // Generate derivations
    let output1 = Command::new("nix-instantiate")
        .arg(tests_dir.join(file1))
        .output()
        .unwrap_or_else(|_| panic!("Failed to instantiate {file1}"));

    let output2 = Command::new("nix-instantiate")
        .arg(tests_dir.join(file2))
        .output()
        .unwrap_or_else(|_| panic!("Failed to instantiate {file2}"));

    let drv1 = String::from_utf8_lossy(&output1.stdout).trim().to_string();
    let drv2 = String::from_utf8_lossy(&output2.stdout).trim().to_string();

    // Run nix-diff with NO_COLOR to get consistent output
    let output = Command::new(env!("CARGO_BIN_EXE_nix-diff"))
        .args([&drv1, &drv2])
        .env("NO_COLOR", "1")
        .output()
        .expect("Failed to run nix-diff");

    assert!(
        output.status.success(),
        "nix-diff failed with stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );

    String::from_utf8_lossy(&output.stdout).to_string()
}

#[test]
fn test_hello_diff_snapshot() {
    let output = run_nix_diff("hello-flake-v1/default.nix", "hello-flake-v2/default.nix");
    assert_snapshot!(output);
}

#[test]
fn test_identical_derivations() {
    let output = run_nix_diff("hello-flake-v1/default.nix", "hello-flake-v1/default.nix");
    assert_snapshot!(output);
}

#[test]
fn test_hello_diff_with_context() {
    let tests_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");

    // Generate derivations
    let output1 = Command::new("nix-instantiate")
        .arg(tests_dir.join("hello-flake-v1/default.nix"))
        .output()
        .expect("Failed to instantiate hello-flake-v1/default.nix");

    let output2 = Command::new("nix-instantiate")
        .arg(tests_dir.join("hello-flake-v2/default.nix"))
        .output()
        .expect("Failed to instantiate hello-flake-v2/default.nix");

    let drv1 = String::from_utf8_lossy(&output1.stdout).trim().to_string();
    let drv2 = String::from_utf8_lossy(&output2.stdout).trim().to_string();

    // Run with different context settings
    let output = Command::new(env!("CARGO_BIN_EXE_nix-diff"))
        .args(["--context", "5", &drv1, &drv2])
        .env("NO_COLOR", "1")
        .output()
        .expect("Failed to run nix-diff");

    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    assert_snapshot!(stdout);
}

#[test]
fn test_word_diff_orientation() {
    let tests_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests");

    // Generate derivations
    let output1 = Command::new("nix-instantiate")
        .arg(tests_dir.join("hello-flake-v1/default.nix"))
        .output()
        .expect("Failed to instantiate hello-flake-v1/default.nix");

    let output2 = Command::new("nix-instantiate")
        .arg(tests_dir.join("hello-flake-v2/default.nix"))
        .output()
        .expect("Failed to instantiate hello-flake-v2/default.nix");

    let drv1 = String::from_utf8_lossy(&output1.stdout).trim().to_string();
    let drv2 = String::from_utf8_lossy(&output2.stdout).trim().to_string();

    // Run with word orientation
    let output = Command::new(env!("CARGO_BIN_EXE_nix-diff"))
        .args(["--orientation", "word", &drv1, &drv2])
        .env("NO_COLOR", "1")
        .output()
        .expect("Failed to run nix-diff");

    assert!(output.status.success());
    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    assert_snapshot!(stdout);
}
