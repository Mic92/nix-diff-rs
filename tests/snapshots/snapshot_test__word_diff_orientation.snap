---
source: tests/snapshot_test.rs
assertion_line: 104
expression: stdout
---
Outputs:
  Output 'out':
    Path:
      - /nix/store/k40bcpw9yrc3cvdr14461m6zs4b2zh7r-hello-v1
      + /nix/store/2apzjc8km4jy9rr055slz7r2bk4s5b11-hello-v2
Arguments:
  Argument 0:
          mkdir -p $out/bin
          cat > $out/bin/hello << 'EOF'
      #!/bin/sh
    - echo "Hello, World! v1"
    + echo "Hello, World! v2"
    + echo "Now with more features!"
      EOF
          chmod +x $out/bin/hello
          
          # Reference dependencies
    -     ln -s /nix/store/zxnf0w2zsjlmyvy503x20dkx9gapfgf3-dep1/bin/dep1 $out/bin/
    -     ln -s /nix/store/cd5mi809khf7wgwij55fj6797vg4zdrb-dep2/share $out/
    +     ln -s /nix/store/pppj168197266j0v77p4sam08g4pgj2p-dep1/bin/dep1 $out/bin/
    +     ln -s /nix/store/ykzvdbr4369678d6nb2kcnzn2bwpsb64-dep2/share $out/
Input derivations:
  - /nix/store/c7ddcxqvcx8vkfqhwqp4vpdj34nr9bll-dep2.drv
  - /nix/store/yg45f58h74lj007n8nyl6v17v0sb2z4i-dep1.drv
  + /nix/store/k886l86g6n3q1cgl5sw1apav22hlsaax-dep1.drv
  + /nix/store/wpqqrnhn872z6vj1r8jbga98yf5c1kjg-dep2.drv
Environment:
  description:
    - A simple hello world program v1
    + A simple hello world program v2 with improvements
  name:
    - hello-v1
    + hello-v2
  newFeature:
    + true
  out:
    - /nix/store/k40bcpw9yrc3cvdr14461m6zs4b2zh7r-hello-v1
    + /nix/store/2apzjc8km4jy9rr055slz7r2bk4s5b11-hello-v2
  version:
    - 1.0
    + 2.0
